### PR TITLE
feat: Add table-header component

### PR DESCRIPTION
**Pull request by eKoopmans**
This PR adds the `d2l-table-header`, which is equivalent to the existing `d2l-list-header` but for tables. Changes include:

- creating the table-header itself, and adding a slot for it into table-wrapper
- extracting how selection-header renders its selection pieces, so that table-header can omit the `d2l-select-all`
- adding the table-header into the table demos, readme, and tests

Rally: https://rally1.rallydev.com/#/?detail=/userstory/656749634867&fdp=true

### CODE REVIEW

File: components/selection/selection-header.js
Comment (eKoopmans) [inline components/selection/selection-header.js:88]: These styles were tightened so that the spacing is still correct when the select-all is omitted (for the table-header).

File: components/table/demo/table-test.js
Comment (eKoopmans) [inline components/table/demo/table-test.js:20]: Changing this to a function means that each table on the demo can retrieve a fresh, independent copy of the data, so their selected state isn't linked.

Also changed the default selected value of two items to demo the selected behaviour more easily (borrowed from @dbatiste's draft PR).

File: components/table/demo/table-test.js
Comment (eKoopmans) [inline components/table/demo/table-test.js:148]: These `scope` changes aren't strictly necessary, just good hygiene (borrowed from the draft PR that was adding support for sticky + scroll-wrapper).

File: components/table/README.md
Comment (eKoopmans) [inline components/table/README.md:331]: This is fairly basic "checked" logic for the readme - I could instead model the more standard pattern of something like `this._data = [ {selected: true, ...}, ... ]`, but I thought that was maybe overkill here.

File: components/table/test/table.test.js
Comment (eKoopmans) [inline components/table/test/table.test.js:44]: @svanherk I've added this in to match the testing in your PR.

My plan is to actually leave this as commented for this PR, and open another PR afterward to uncomment this test and add langterms etc. That way this PR won't get even bigger 😄.

**Review by eKoopmans — COMMENTED** [review]
This PR will have a merge conflict once #3040 is merged, because we're both tinkering with the internals of `d2l-selection-header`.

File: components/table/README.md
Comment (dbatiste) [inline components/table/README.md:351]: I think in practice the linter/analyzer might complain about calling functions in handlers defined in templates. I think it's a perf thing to avoid the recreation of functions when the templates re-render. To avoid it, we'd normally use the same handler for all the `d2l-selection-input`, and use `e.target.key` to do the necessary look-up.

**Review by oelhanafey — COMMENTED** [review]
The demo page seems to be working locally so I don't think it is anything with your code here, but it looks like the table demo page on the PR site is not working properly:
https://pr-3051-brightspace-ui-core.d2l.dev/components/table/demo/table.html

File: components/selection/selection-header.js
Comment (eKoopmans) [reply components/selection/selection-header.js:166]: My plan was to keep this as the default implementation. The `d2l-selection-header` currently works as-is, making it empty/abstract would break that.

I know there was some debate about whether it's desirable or not for `d2l-selection-header` to exist standalone, though I can say that it's used directly in our selection demos (on the readme and the internal dev demos). It was a perfect drop-in replacement for the custom-made "selection header" we had in the examples previously.

File: components/table/test/table.visual-diff.js
Comment (eKoopmans) [reply components/table/test/table.visual-diff.js:136]: This one I'm not sure about - I just followed the pattern that was already there. The sticky tests could potentially cause conflicts, though it doesn't seem likely.

@dlockhart the existing tests have your name on it (though it's just "migrate to Lit"), do you know more about why these are selectively shown/hidden?

Comment (dlockhart) [reply components/table/test/table.visual-diff.js:136]: Hmm unfortunately I really don't recall... yeah possibly an artifact from before the migration to Lit and/or sticky stuff.

**Review by oelhanafey — COMMENTED** [review]
LGTM! But I'll let someone else have a look and approve it

File: components/table/table-header.js
Comment (dlockhart) [inline components/table/table-header.js:6]: Don't think this slot description should be here? Looks like it was copied from `selection-header.js` to list as well.

File: components/table/table-header.js
Comment (dlockhart) [inline components/table/table-header.js:16]: `d2l-selection-select-all-pages.js` should likely be explicitly imported, even though we're getting it for free from `selection-header.js`.

**Comment by eKoopmans** [comment]
Ping - any additional feedback on this one?

**Review by dlockhart — APPROVED** [review]
Looks excellent to me

**Pull request merged**
